### ISSUE TITLE
`react-x/no-unused-state` should distinguish between effect orchestration and synchronous setState anti-patterns

### ISSUE THREAD

**Issue by rarenatoe** [Rel1cx/eslint-react#1749]
## Description

The `react-x/no-unused-state` rule (restored in v5.7.0) flags all state variables used exclusively in effects as "unused", but doesn't distinguish between:
- ✅ **Valid pattern**: Async setState in callbacks responding to external systems
- ❌ **Anti-pattern**: Synchronous setState in effect body (creates cascading renders)

## Problem

The rule currently flags both patterns identically, making it difficult to identify actual bugs vs legitimate async state synchronization.

## Valid Use Cases (Should NOT Be Flagged)

State set **asynchronously in callbacks** responding to external systems is the **intended React pattern** per [React docs](https://react.dev/learn/you-might-not-need-an-effect).

### Example 1: External Service State
```typescript
const [serviceReady, setServiceReady] = useState(false)

useEffect(() => {
  if (!serviceReady && hasPermission) {
    void ExternalService.initialize(config)
      .then(() => setServiceReady(true)) // ✅ Async callback
  }
}, [serviceReady, hasPermission, config])

useEffect(() => {
  if (!serviceReady) return
  // Use the service only after it's ready
  ExternalService.start()
}, [serviceReady])
```

**Why valid:** setState responds to external API. Second effect waits for initialization.

### Example 2: Async Data Loading
```typescript
const [data, setData] = useState('')

useEffect(() => {
  const load = async () => {
    const value = await fetchFromStorage()
    setData(value) // ✅ Async callback
  }
  void load()
}, [])

useEffect(() => {
  if (data) {
    doSomethingWithData(data)
  }
}, [data])
```

**Why valid:** Cannot load synchronously. Second effect waits for data.

### Example 3: Multi-Step Flow
```typescript
const [authenticated, setAuthenticated] = useState(false)

const login = async () => {
  const success = await authService.login()
  setAuthenticated(success) // ✅ Triggers post-auth effect
}

useEffect(() => {
  if (authenticated) {
    void performPostLoginSetup()
  }
}, [authenticated])
```

**Why valid:** Auth state drives subsequent actions. Common flow pattern.

### ❌ Anti-pattern: Synchronous setState in Effect
```typescript
const [flag, setFlag] = useState(false)

useEffect(() => {
  if (condition && !flag) {
    setFlag(true) // ❌ Synchronous in effect body
    doSomething()
  }
}, [condition, flag]) // ← Cascading renders
```

**Why wrong:** 
- Synchronous setState in effect body
- Effect depends on state it modifies
- Creates cascading renders

**Fix:** Use `useRef` instead:
```typescript
const flagRef = useRef(false)

useEffect(() => {
  if (condition && !flagRef.current) {
    flagRef.current = true
    doSomething()
  }
}, [condition])
```

### Option 1: Detect Synchronous vs Async
- **Error**: State set synchronously in effect body
- **Warn/Off**: State only set in async callbacks

### Option 2: Configuration
```typescript
'react-x/no-unused-state': ['warn', {
  reportSynchronousSetState: 'error', // Actual bugs
  reportAsyncSetState: 'off',         // Valid pattern
}]
```

### Option 3: Better Messages
```
❌ Current: "State variable 'x' is defined but never used, or only used in effects"

✅ Better: 
  - "State 'x' set synchronously in effect - use useRef instead"
  - "State 'x' only used in async callbacks - verify this is intentional"
```

## Why This Matters

**Current behavior creates false equivalence:**
- Valid async external state synchronization ← flagged as "unused"
- Actual bugs (synchronous setState) ← also flagged as "unused"

Developers cannot distinguish between "fix this bug" and "this is fine."

### Platform and versions

```Text
**Environment:**
- eslint-plugin-react-x: 5.7.1
- React: 18.x
- React Native: 0.81.x
- node: 24.15.0
```

**Comment by silverwind** [comment]
*This comment was written by Claude, an AI assistant.*

Adding a variant your existing examples don't quite cover: state used purely as a dep-array tripwire, never read inside the effect body.

```tsx
function ReloadButton() {
  const [reloadKey, setReloadKey] = useState(0);

  useEffect(() => {
    refetch();                     // does NOT read reloadKey
  }, [reloadKey]);                 // dep-array read only

  return <button onClick={() => setReloadKey(k => k + 1)}>Reload</button>;
}
```

A ref can't substitute — dep arrays compare by reference equality and `ref.current` mutations don't re-trigger effects.
